### PR TITLE
Implicit casts must conform with data precedence

### PR DESCRIPTION
**Pull request by jeeminso**
## Summary of the changes / Why this improves CrateDB
Follows https://github.com/crate/crate/pull/16622#discussion_r1762540353.

CrateDB only supports `like any (string, array(string))` variant but array(double) is implicitly converted to array(string)(lower precedence) and successfully invokes the `(string, array(string))` version. Implicit conversions must follow data precedence.

## Checklist

 - [ ] Added an entry in the latest `docs/appendices/release-notes/<x.y.0>.rst` for user facing changes
 - [ ] Updated documentation & `sql_features` table for user facing changes
 - [ ] Touched code is covered by tests
 - [ ] This does not contain breaking changes, or if it does:
    - It is released within a major release
    - It is recorded in the latest `docs/appendices/release-notes/<x.y.0>.rst`
    - It was marked as deprecated in an earlier release if possible
    - You've thought about the consequences and other components are adapted
      (E.g. AdminUI)

### CODE REVIEW

File: server/src/main/java/io/crate/types/DataType.java [hunk 1]
```diff
@@ -71,9 +71,9 @@ public enum Precedence {
         BOOLEAN,
         SHORT,
         TIMETZ,
+        REGCLASS,
         INTEGER,
```
Comment (jeeminso) [inline server/src/main/java/io/crate/types/DataType.java:75]:
```
ERROR: Unknown function: pg_catalog.pg_get_expr(d.adbin, d.adrelid), no overload found for matching argument types: (text, regclass). Possible candidates: pg_catalog.pg_get_expr(text, integer):text, pg_catalog.pg_get_expr(text, integer, boolean):text
```
There are hundreds of this failures, need to allow implicit cast from REGCLASS to INTEGER.

**Comment by mfussenegger** [comment]
Could you please add some problem explanations to the PR description?

E.g.: in the format of Problem: <description of user observable behavior> solution: ...


Changing cast semantics is risky, I'd rather not backport this to 5.8 unless there is a substantial issue with it.

**Comment by jeeminso** [comment]
> Could you please add some problem explanations to the PR description?

Sorry added.

**Comment by mfussenegger** [comment]
> CrateDB only supports like any (string, array(string)) variant but array(double) is implicitly converted to array(string)(lower precedence) and successfully invokes the (string, array(string)) version. Implicit conversions must follow data precedence.

This is because `like any` only has a `string` overload. 
The logic in the function resolution is to cast to higher precedence, but only if there is a signature match. It will do downcasts as fallback.

This is working as expected.

**Comment by jeeminso** [comment]
Thank you for clarifying @mfussenegger , closing.

**Pull request closed without merging**